src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LanguageServerProjectLoader.cs

```diff
@@ -273,4 +362,80 @@ async Task LogDiagnosticsAsync(ImmutableArray<DiagnosticLogItem> diagnosticLogIt
             await toastErrorReporter.ReportErrorAsync(worstLspMessageKind, message, cancellationToken);
         }
     }
+
+    protected async ValueTask TrackPrimordialOnlyProjectAsync(string projectPath, ProjectId primordialProjectId)
+    {
+        using (await _gate.DisposableWaitAsync(CancellationToken.None))
+        {
+            // If project is already tracked, no need to do any further work.
+            if (_loadedProjects.ContainsKey(projectPath))
+            {
+                return;
+            }
+
+            _loadedProjects.Add(projectPath, new ProjectLoadState.Primordial(primordialProjectId));
+        }
+    }
+
+    protected async ValueTask BeginLoadingProjectWithPrimordialAsync(string projectPath, ProjectId primordialProjectId)
+    {
+        using (await _gate.DisposableWaitAsync(CancellationToken.None))
+        {
+            // If project has already begun loading, no need to do any further work.
+            if (_loadedProjects.ContainsKey(projectPath))
+            {
+                return;
+            }
+
+            _loadedProjects.Add(projectPath, new ProjectLoadState.Primordial(primordialProjectId));
+            _projectsToReload.AddWork(new ProjectToLoad(projectPath, ProjectGuid: null, ReportTelemetry: true));
```

Comment:
Consider merging this with the prior method just to take a boolean that does this line or not.